app/modules/site_settings/helpers.py

```diff
@@ -276,6 +277,51 @@ def get_definition(cls, cls_name, guid):
                 return defn
         return None
 
+    @classmethod
+    # replace=False will silently fail if already exists
+    def add_definition(cls, class_name, guid, defn, replace=False):
+        from .models import SiteSetting
+
+        # bad class_name will get raise HoustonException
+        data = SiteSetting.get_value(f'site.custom.customFields.{class_name}')
+        if not data or not isinstance(data.get('definitions'), list):
+            data = {'definitions': []}
+        found = False
+        for i in range(len(data['definitions'])):
+            if guid == data['definitions'][i].get('id'):
+                if replace:
+                    found = True
+                    data['definitions'][i] = defn
+                else:
+                    return  # exists - no update!
+        if not found:
+            data['definitions'].append(defn)
+        AuditLog.audit_log(log, f'add_definition added {guid} to {class_name}')
+        SiteSetting.set(f'site.custom.customFields.{class_name}', data=data)
+
+    # WARNING: this does no safety check (_drop_data etc), so really other code that
+    #   wraps this should be used, e.g. patch_remove()
+    @classmethod
+    def remove_definition(cls, class_name, guid):
+        from .models import SiteSetting
+
+        # bad class_name will get raise HoustonException
+        data = SiteSetting.get_value(f'site.custom.customFields.{class_name}')
+        if not data or not isinstance(data.get('definitions'), list):
+            return
+        new_list = []
+        found = False
+        for defn in data['definitions']:
+            if guid == defn.get('id'):
+                found = True
+            else:
+                new_list.append(defn)
+        if found:
+            AuditLog.audit_log(log, f'remove_definition dropped {guid} for {class_name}')
+            SiteSetting.set(
+                f'site.custom.customFields.{class_name}', data={'definitions': new_list}
+            )
+
     # expects cf_defn as from above
     # NOTE this is very very likely incomplete.  the 'type' value in definitions are quite complex.
```

Comment:
Aside, is there a Jira associated with this TODO?

Reply:
DEX-1336 and DEX-1351 are kind both related to this whole mess of just how fine-tuned things can get.  DEX-1362 also happend to catch a couple things mentioned here too, especially depending on if by TODO you mean all the notes, comments, and FIXMEs in this section.